keps/sig-auth/20190916-noderestriction-pods.md

```diff
@@ -0,0 +1,476 @@
+  host namespaces (i.e. a container escape). Low-privilege in this case means the node is not
+  hosting any privileged workloads with permissions to trivially take over the cluster
+  (e.g. unrestricted pod creation, read secrets, etc).
+
+### Goals
+
+- Prevent a compromised node from manipulating controllers to execute confused deputy attacks.
+
+### Non-Goals
+
+- Solving all node isolation issues.
+- Solving all possible man-in-the-middle attacks.
+
+## Proposal
+
+### Node Restrictions
+
+All restrictions will be enforced through the [NodeRestriction][] admission controller. These
+extensions will be guarded by the `NodeRestrictionPods` feature gate.
+
+On update, only the delta will be restricted. In other words, the Kubelet can modify pods with
+restricted labels / owner references as long as it's not modifying (or adding or
+deleting) the restricted entries.
+
+#### Label Restrictions
+
+The Kubelet will be prevented from updating pod labels or creating mirror pods with labels, except
+for whitelisted keys:
+
+- Any labels starting with `unrestricted.node.kubernetes.io/` are allowed.
+
+The Kubelet does not currently label pods, nor are there official label keys that apply to
+pods. However, there are a few labels that are commonly applied to system addons & static pods:
+
+- `component` (used by [kubeadm][kubeadm-labels])
+- `tier` (used by [kubeadm][kubeadm-labels])
+- `k8s-app` (common on [addons][addons-k8s-app])
+
+The `k8s-app` label is used to match controllers for system components, and therefore should be
+explicitly disallowed.
+
+It is not clear how the `tier` and `component` labels are consumed, but I recommend uses be migrated
+to `unrestricted.node.kubernetes.io/component` and `unrestricted.node.kubernetes.io/tier`, rather
+than special casing the existing labels.
+
+[kubeadm-labels]: https://github.com/kubernetes/kubernetes/blob/e682310dcc5d805a408e0073e251d99b8fe5c06d/cmd/kubeadm/app/util/staticpod/utils.go#L60
+[addons-k8s-app]: https://github.com/kubernetes/kubernetes/blob/e682310dcc5d805a408e0073e251d99b8fe5c06d/cluster/addons/kube-proxy/kube-proxy-ds.yaml#L23
+
+#### OwnerReferences
+
+OwnerReferences cannot be updated through the `pod/status` subresource, but they can be set on
+mirror pods. With the new restrictions, mirror pods are only allowed a single owner reference (or
+none), and it must refer to the node:
+
+```go
+ metav1.OwnerReference{
+  APIVersion: "v1"
+  Kind: "Node"
+  Name: node.Name
+  UID:  node.UID
+  Controller: true
+  BlockOwnerDeletion: false
+}
+```
+
+The node owner reference will eventually be required, but due to apiserver-node version skew, this
+must happen at least 2 releases after `MirrorPodMetadata` graduates to beta.
+
+### Kubelet changes
+
+To comply with and ease the rollout of the modified restrictions, Kubelet will make some
+modifications to mirror pods. The Kubelet changes will be guarded by a separate
+`MirrorPodMetadata` feature gate.
```

Comment:
suggest a name that makes it clear this gates kubelet rewrite behavior, like `MirrorPodMetadataKubeletFilter`

Reply:
Ok with `MirrorPodMetadataFilter`?